compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12845,195 +12845,303 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    }
 
 
+
+
+// Overflow/Underflow handling:
+// Must detect overflow without causing overflow:
+// Equation: add operation: x + Const1 < Const2 --> x < Const2 - Const1
+//          sub operation: x - Const1 < Const2 --> x < Const2 + Const1
+
+//parameters are unsigned as signed overflow is undefined behaviour.
+//Therefore to check if Leftmost bit is set, we check if const > (unsigned)max
+//Since negative values are greater than positive values,take the two's compliment for calculations
+
+//Signed Overflow occurs for both operations, Unsigned only through a sub operation
+bool isNotOverflow(TR::ILOpCodes op, uint64_t oldConst1, uint64_t oldConst2, uint64_t max){
+
+   bool isAddOp = (op == TR::iadd || op == TR::ladd || op == TR::sadd || op == TR::badd);
+   bool isSubOp = (op == TR::isub || op == TR::lsub || op == TR::ssub || op == TR::bsub);
+
+   return (isAddOp && !(oldConst1 > max && oldConst2 <= max && (oldConst2 > max - ((~oldConst1+1) & max ))))
+            || (isSubOp && !(oldConst1 <= max && oldConst2 <= max && (oldConst2 > max - oldConst1)));
```

Comment:
This kind of code is dangerous. The old code for checking signed overflow was correct. 

```
(isAddOp && !(oldConst1 < 0 && (oldConst2 > signedMax + oldConst1)))
```

Note that the `oldConst1 < 0` part guarantees that `signedMax + oldConst1 < signedMax`, meaning that no overflow happens and the behaviour is well defined. This is also the more generally accepted way of checking for signed overflow because it doesn't use type conversions (which don't always work the way you expect them to).

(For the record, I've already checked the code for undefined behaviour.)

Reply:
Thanks for that @Leonardo2718 I had mis-read the condition and the limitation on the const was what I missed. Sorry for the churn @kobinau

Reply:
alright I returned to the original code. it now consolidates both unsigned and signed cases into one overflow and one underflow case